### ISSUE TITLE
Where to put Audit.Core.Configuration.Setup()

### ISSUE THREAD

**Issue by mrlife** [thepirat000/Audit.NET#378]
Where can/should this get called? I could not find in the readme the location to call `Audit.Core.Configuration.Setup()`.

I first tried calling `Audit.Core.Configuration.Setup()` in the database context `OnModelCreating()`, and the audit table was created. However, there were some null errors when trying to save to the audit table.

Second, I tried tried calling `Audit.Core.Configuration.Setup()` in the database context `OnConfiguring()`, since I saw it in [another issue](https://github.com/thepirat000/Audit.NET/issues/308). That resulted in the audit table being deleted, so that doesn't seem like the right place.

> Your configuration call to `Audit.Core.Configuration.Setup()` should be done only once, at the startup of your application.

_Originally posted by @thepirat000 in https://github.com/thepirat000/Audit.NET/issues/247#issuecomment-539686730_

**Comment by thepirat000** [comment]
What kind of application? If it's asp.net you could do it on the strartup.
Otherwise in any place that is run before any audited action takes places, and that runs only once of course...

On the static constructor of your dbcontext is another choice, that will run the first time a dbcontext is created and only once.

**Comment by mrlife** [comment]
@thepirat000 I'm on ASP.NET Core 5.0.4. I have an EF query that needs to run to look up the user information, so the static constructor or startup.cs aren't great options in this case. However, as a test, I removed the EF query and placed the `Setup()` call in the static constructor, and the audit table e.g. Audit_Order is not created in the database.

I have `Audit.EntityFramework.Core` v16.5.4 installed on a .NET/EF Core 5.0.4 project. I am inheriting `Audit.EntityFramework.AuditDbContext` from the database context. And I have the below `Setup()` call. Everything compiles fine, but the `Audit_MyModel` table isn't created when I run a migration. Any thoughts on what I'm doing wrong?

```
Audit.Core.Configuration.Setup()
    .UseEntityFramework(ef => ef
        .AuditTypeExplicitMapper(m => m
            .Map<MyModel, Audit_MyModel>()
            .AuditEntityAction<AuditBase>((evt, entry, auditEntity) =>
            {
                auditEntity.AuditDate = DateTime.Now;
                auditEntity.AuditAction = entry.Action; // Insert, Update, Delete
            })
        )
    );
```

**Comment by thepirat000** [comment]
Audit table creation is up tu you, the library will not create the tables.

Why do you say the startup is not a great option to put the configuration?

**Comment by mrlife** [comment]
> Audit table creation is up to you, the library will not create the tables.

Do users of the library typically do this with `DbSet`s?

> Why do you say the startup is not a great option to put the configuration?

I would need to access user claims through DI and to look up information about the user, i.e.

```
Audit.Core.Configuration.Setup()
    .UseEntityFramework(ef => ef
        .AuditTypeExplicitMapper(m => m
            .Map<Order, Audit_Order>()
            .Map<OrderItem, Audit_OrderItem>()
            .AuditEntityAction<IAudit>((evt, entry, auditEntity) =>
            {
                // access to claims through DI and to query a DbSet here
            })
        )
    );
```

**Comment by thepirat000** [comment]
For the DI part, you should be able to access any of your registered services, or the `IServiceProvider` as a parameter and then use it on the delegate, for example:

```c#
public void Configure(IApplicationBuilder app, IServiceProvider svcProvider)
{

    Audit.Core.Configuration.Setup()
        .UseEntityFramework(ef => ef
            .AuditTypeExplicitMapper(m => m
                .Map<Order, Audit_Order>()
                .Map<OrderItem, Audit_OrderItem>()
                .AuditEntityAction<IAudit>((evt, entry, auditEntity) =>
                {
                    var svc = svcProvider.GetRequiredService<IYourService>();
                    // ...
                })
            )
        );
}
```

**Comment by mrlife** [comment]
_@thepirat000 said:_
> Is the DbSet you need to access on the same DbContext that is being audited?

Yes, that it awesome.

Thank you for offering this kind of support for your library. It should be a wonderful aspect of my project after I get it set up.

As for the location of `Audit.Core.Configuration.Setup()`:

After moving `Audit.Core.Configuration.Setup()` to `Startup.Configure()`, I've realized that since getting authentication state (i.e. with built-in, _**user-scoped**_ `AuthenticationStateProvider` service in Blazor), I won't be able to get the userid out of the claims if it's called in Startup.cs. For posterity, the error is:

> GetAuthenticationStateAsync was called before SetAuthenticationState.

When I had `Audit.Core.Configuration.Setup()` in the context's static constructor, I was getting the following error. This was after adding the `Audit_Order` as a `DbSet`. Any thoughts on this one?

> The instance of entity type 'Audit_Order' cannot be tracked because another instance with the key value '{Id: 1}' is already being tracked. When attaching existing entities, ensure that only one entity instance with a given key value is attached.

**Comment by mrlife** [comment]
To help illustrate, here is a [sample app](https://github.com/thepirat000/Audit.NET/files/6151661/BlazorServerDbContextExample.zip). It's the [official Blazor with EF Core sample](https://docs.microsoft.com/en-us/aspnet/core/blazor/blazor-server-ef-core?view=aspnetcore-5.0) with the following modifications to add in this library.

I added the package to the `.csproj`, added `AuditDbContext` for inheritance, created `Audit_Contact`, and added the `Setup()` call to the context's static constructor.

After running the app and deleting one of the records, this is the error:

> UNIQUE constraint failed: Audit_Contacts.Id

**Comment by thepirat000** [comment]
You should remove any unique index based on the audited table ID on the audit table, since the audit table will potentially contain multiple rows for the same ID. I guess that's the problem

**Comment by mrlife** [comment]
Thank you, I fixed by adding an explicit key property to the `Audit_*` model classes so that the `Id` matched up with the audited model and EF Core uses the explicit key property for its tracking purposes.

e.g.
```
public class MyModel
{
    public int Id { get; set; }
    public string Property { get; set; }
}
 
public class Audit_MyModel : IAudit
{
    [Key]
    public int AuditId { get; set; }
    public int Id { get; set; }
    public string Property { get; set; }
 
    public string AuditAction { get; set; }
    public DateTime AuditDate { get; set; }
}
```

I think your library is wonderful and appreciate your work to create and maintain it.
